lib/membrane_file/sink.ex

```diff
@@ -1,48 +1,116 @@
+  field of the event to `true` to start inserting new buffers without overwriting previous ones.
+  Please note, that inserting requires rewriting the file, what negatively impacts performance.
+  """
   use Membrane.Sink
-  alias Membrane.Buffer
-  alias Membrane.File.CommonFile
-
   import Mockery.Macro
 
-  def_options location: [type: :string, description: "Path to the file"]
+  alias Membrane.File.{CommonFile, Error, SeekEvent}
 
-  def_input_pad :input, demand_unit: :buffers, caps: :any
+  def_options location: [
+                spec: String.t(),
+                description: "Path to the file"
+              ]
 
-  # Private API
+  def_input_pad :input, demand_unit: :buffers, caps: :any
 
   @impl true
   def handle_init(%__MODULE__{location: location}) do
     {:ok,
      %{
        location: location,
-       fd: nil
+       temp_location: location <> ".tmp",
+       fd: nil,
+       temp_fd: nil
      }}
   end
 
   @impl true
-  def handle_stopped_to_prepared(_ctx, state), do: mockable(CommonFile).open(:write, state)
+  def handle_stopped_to_prepared(_ctx, %{location: location} = state) do
+    case mockable(CommonFile).open(location, [:read, :write]) do
+      {:ok, fd} -> {:ok, %{state | fd: fd}}
+      error -> Error.wrap(error, :open, state)
+    end
+  end
 
   @impl true
   def handle_prepared_to_playing(_ctx, state) do
     {{:ok, demand: :input}, state}
   end
 
   @impl true
-  def handle_write(:input, %Buffer{payload: payload}, _ctx, %{fd: fd} = state) do
-    bin_payload = Membrane.Payload.to_binary(payload)
+  def handle_write(:input, buffer, _ctx, %{fd: fd} = state) do
+    case mockable(CommonFile).write(fd, buffer) do
+      :ok -> {{:ok, demand: :input}, state}
+      error -> Error.wrap(error, :write, state)
+    end
+  end
 
-    with :ok <- mockable(CommonFile).binwrite(fd, bin_payload) do
-      {{:ok, demand: :input}, state}
+  @impl true
+  def handle_event(:input, %SeekEvent{insert?: insert?, position: position}, _ctx, state) do
+    if insert?, do: split_file(position, state), else: seek_file(position, state)
+  end
+
+  def handle_event(pad, event, ctx, state), do: super(pad, event, ctx, state)
+
+  @impl true
+  def handle_prepared_to_stopped(_ctx, %{fd: fd} = state) do
+    with {:ok, state} <- maybe_merge_temporary(state),
+         :ok <- mockable(CommonFile).close(fd) do
+      {:ok, %{state | fd: nil}}
     else
-      {:error, reason} -> {{:error, {:write, reason}}, state}
+      error -> Error.wrap(error, :close, state)
     end
   end
 
-  @impl true
-  def handle_prepared_to_stopped(_ctx, state), do: mockable(CommonFile).close(state)
+  defp seek_file(position, %{fd: fd} = state) do
+    with {:ok, state} <- maybe_merge_temporary(state),
+         {:ok, _position} <- mockable(CommonFile).seek(fd, position) do
+      {:ok, state}
+    else
+      error -> Error.wrap(error, :seek_file, state)
+    end
+  end
+
+  defp split_file(position, %{fd: fd} = state) do
```

Comment:
Does it work if you already had an opened temporary file? From the code, I guess you don't support several splits before merging them (if that is intentional then add a warning and a TODO).

Reply:
It works, but temporary needs to be merged first. Then the main file is splitted again 😛 
Added `# TODO` about improving performance for multiple insertions in the future

Reply:
Ahh, I missed the `split_file` calling the `seek_file` :)